src/solvers/dgsem_p4est/dg_parallel.jl

```diff
@@ -0,0 +1,296 @@
+
+# Function barrier for type stability
+function init_neighbor_rank_connectivity_iter_face_inner(info, user_data)
+  @unpack interfaces, interface_id, global_interface_ids, neighbor_ranks_interface, mesh = user_data
+
+  # Get the global interface ids and neighbor rank if current face belongs to an MPI interface
+  if info.sides.elem_count == 2 # MPI interfaces have two neighboring elements
+    # Extract surface data
+    sides = (unsafe_load_side(info, 1), unsafe_load_side(info, 2))
+
+    if sides[1].is_hanging == 0 && sides[2].is_hanging == 0 # No hanging nodes for MPI interfaces
+      if sides[1].is.full.is_ghost == 1
+        remote_side = 1
+        local_side = 2
+      elseif sides[2].is.full.is_ghost == 1
+        remote_side = 2
+        local_side = 1
+      else # both sides are on this rank -> skip since it's a regular interface
+        return nothing
+      end
+
+      # Sanity check, current face should belong to current MPI interface
+      local_tree = unsafe_load_tree(mesh.p4est, sides[local_side].treeid + 1) # one-based indexing
+      local_quad_id = local_tree.quadrants_offset + sides[local_side].is.full.quadid
+      @assert interfaces.local_element_ids[interface_id] == local_quad_id + 1 # one-based indexing
+
+      # Get neighbor ID from ghost layer
+      proc_offsets = unsafe_wrap(Array, info.ghost_layer.proc_offsets, mpi_nranks() + 1)
+      ghost_id = sides[remote_side].is.full.quadid # indexes the ghost layer, 0-based
+      neighbor_rank = findfirst(r -> proc_offsets[r] <= ghost_id < proc_offsets[r+1],
+                                1:mpi_nranks()) - 1 # mpi ranks are 0-based
+      neighbor_ranks_interface[interface_id] = neighbor_rank
+
+      # Global interface id is the globally unique quadrant id of the quadrant on the primary
+      # side (1) multiplied by the number of faces per quadrant plus face
+      if local_side == 1
+        offset = unsafe_load(mesh.p4est.global_first_quadrant, mpi_rank() + 1) # one-based indexing
+        primary_quad_id = offset + local_quad_id
+      else
+        offset = unsafe_load(mesh.p4est.global_first_quadrant, neighbor_rank + 1) # one-based indexing
+        primary_quad_id = offset + sides[1].is.full.quad.p.piggy3.local_num
+      end
+      global_interface_id = 2 * ndims(mesh) * primary_quad_id + sides[1].face
+      global_interface_ids[interface_id] = global_interface_id
+
+      user_data.interface_id += 1
+    end
+  end
+
+  return nothing
+end
+
+# TODO: ::Any is a hack. This method is already defined in the TreeMesh
+# code. It uses `mpi_neighbor_mortars` as an argument but not uEltype, thus the number
+# of arguments would be the same without ::Any which would break precompilation.
+function init_mpi_data_structures(mpi_neighbor_interfaces, n_dims, nvars, n_nodes, uEltype, ::Any)
```

Comment:
@ranocha What's your take on this? The purpose and output of the two function implementations is comparable, but they require considerably different input. What would be the cleanest way to solve this: Rename the functions (since they are functionally different) or dispatch on an argument?

Reply:
I don't have a strong opinion on this. If it makes sense to unify the function names, go for it. If they are sufficiently different, choose different names. They are not part of the public API anyway

Reply:
OK.  I slightly lean towards making this a different function name, which imho would be cleaner than having an unused argument attached just for dispatch purposes. @lchristm I'll leave it up to you ;-)

Reply:
I think once the `ParallelP4estMesh` supports non-conforming meshes and the `ParallelTreeMesh` supports custom `uEltype`s, the two meshes could share a lot of the MPI-related code, in particular this function here. So my suggestion would be to leave it as is for now and use a single function for both meshes once the missing features have been implemented.

Reply:
OK. Maybe you can add a reference to this convo to the comment such that it can be easily recovered in the future. Afterwards, feel free to resolve this convo.

Reply:
Done in 1981be8.